Kcena restaurant inquiry

Kwakcena opened the pull request:
현재 레스토랑의 지역 목록을 불러오는 부분까지 완료했습니다. 나머지도 곧 완료 하겠습니다.

Kwakcena committed:
Init commit

Kwakcena committed:
test: 레스토랑 지역 목록 보여주기

레스토랑의 지역 목록을 보여주는 테스트 작성.

Kwakcena committed:
test(green): 레스토랑 지역 목록 보여주기

하드코딩으로 통과시킴

Kwakcena committed:
test(Refactor): 레스토랑 지역 목록 컴포넌트 분리

레스토랑 지역 목록을 보여주는 컴포넌트를 분리

Kwakcena committed:
test(red): 컨테이너 내부에서 UI 관련 로직 분리

레스토랑 지역 목록을 받아 보여주는 컴포넌트로 RestaurantRegions
컴포넌트를 만들어 분리했다. 현재는 해당 컴포넌트가 존재하지 않아서
테스트는 실패한다.

Kwakcena committed:
test(red): 레스토랑 지역 목록 보여주기 테스트

컴포넌트 분리 후 테스트

Kwakcena committed:
test(green): RestaurantRegions 컴포넌트 생성

지역 목록 받아서 보여주기 테스트 통과

Kwakcena committed:
test(Refactor): fixtures 생성

테스트용 데이터 생성

Kwakcena committed:
Refactor: 지역 목록은 버튼으로 구성되어 있음

버튼으로 구성된 지역 목록으로 수정

Kwakcena committed:
Refactor: test용 데이터 사용

Kwakcena committed:
test(red): useSelector의 사용

useSelector를 이용해서 지역 목록을 불러오길 원함

Kwakcena committed:
test(red): useSelector로 regions 가져오기

useSelector로 regions를 가져오기 위해서는 reducer가 필요하다.

Kwakcena committed:
feat: reducer를 사용하기 위한 코드 추가

reducer.js, store.js 추가

Kwakcena committed:
test(green): 가짜 useSelector 만들어 통과

가짜 useSelector를 만들어 regions를 가져와 통과시킨다.

Kwakcena committed:
test(green): App 컴포넌트 useSelector mock 사용

App 컴포넌트도 테스트를 통과하기 위해 useSelector mocking 하여 통과시킴.

Kwakcena committed:
test(red): 서버로부터 데이터를 가져오기를 원함

App 컴포넌트에서 useEffect를 이용해 서버로부터 데이터를 가져오기를
원한다. 이를 구현체에 넣었고 이것이 가능하도록 reducer에서 테스트를
진행해야 한다.

Kwakcena committed:
test(red): 컴포넌트 실행 시 지역 정보 가져오기

일단 서버로부터가 아닌 fixtures를 잘 가져오는지 테스트 한다.

App 컴포넌트가 실행되면 useEffect에 의해 loadRestaurantRegions 가
실행되고 내부에 있는 dispatch에 의해 setRegions가 실행될 것이다.

하지만 지금은 setRegions가 없으므로 실패한다.

Kwakcena committed:
test(green): useDispatch mocking

App 컴포넌트가 actions에 setRegions가 없어서 테스트가 실패한 것이 아니라
useDispatch를 사용하고 있기 때문에 이를 mocking 해야 한다.

이에 대해 추가하여 테스트를 통과시킨다.

Kwakcena committed:
test(red) - 레스토랑 지역 목록 set 테스트

레스토랑의 지역 정보를 regions 배열에 set하는 reducer 테스트를 진행.

구현되어있지 않아서 테스트는 실패

Kwakcena committed:
test(green): setRegions 테스트 통과

Kwakcena committed:
test(red): 실제 api로 regions 가져오기

이제 fixtures의 데이터가 아닌 api를 이용하여 가져온 데이터로 dispatch를
해보자.

이를 위해서는 api가 구현되어 있어야 한다.

Kwakcena committed:
test(green): App test 에서 api mocking 하여 통과

App 컴포넌트 테스트에서는 api에 대해 알지 못하므로 이를 mocking 해야
한다.

Kwakcena committed:
feat: 실제 api 기능 구현

api 기능을 실제로 구현하여 지역 목록을 받아온다.

Kwakcena committed:
test(red): 우리가 원하는 기능

thunk를 이용해서 비동기 액션을 만들어 보자.

Kwakcena committed:
thunk 사용 준비

thunk를 사용하기 위해 미들웨어를 등록한다.

Kwakcena committed:
test(green): 비동기 액션 사용

리덕스의 비동기 액션을 만들어 지역 목록을 가져오는데 성공한다.

Kwakcena committed:
test(Refactor): dispatch called 테스트 추가

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,36 @@
+
+describe('App', () => {
+  it('레스토랑 지역 목록을 보여준다.', () => {
+    const dispatch = jest.fn();
+    useDispatch.mockImplementation(() => dispatch);
+    useSelector.mockImplementation((selector) => selector({
+      regions,
+    }));
```
`it`안에는 given when then에서 when과 then에 해당하는 부분이에요. given에 대한 것들은 바깥으로 분리시켜주는게 보기 좋을 것 같아요.

```js
const dispatch = jest.fn();

beforeEach(() => {
  dispatch.mockClear();

  useDispatch.mockImplementation(() => dispatch);
  useSelector.mockImplementation((selector) => selector({ regions }));
});
```

Kwakcena replied:
given when then이 무엇인가요 ?

hannut91 replied:
다음 영상을 참고해주세요.

https://youtu.be/gfTsSBRvdqI

Kwakcena replied:
이 영상 예전에 슬랙에 올라왔을 떄 봣는데 given when then에 대한 언급이 있었나요...?!

hannut91 replied:
아 없군요

hannut91 replied:
https://github.com/ahastudio/til/blob/master/blog/2018/12-08-given-when-then.md 이거네요

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,36 @@
+
+    expect(dispatch).toBeCalledTimes(1);
+
+    expect(getByText('서울')).not.toBeNull();
+    expect(getByText('대전')).not.toBeNull();
+    expect(getByText('대구')).not.toBeNull();
+    expect(getByText('부산')).not.toBeNull();
+    expect(getByText('광주')).not.toBeNull();
+    expect(getByText('강원도')).not.toBeNull();
+    expect(getByText('인천')).not.toBeNull();
```
fixture를 사용하면 더 간단하게 표현할 수 있을 것 같아요

hannut91 commented on src/RestaurantRegions.test.jsx:
```diff
@@ -0,0 +1,23 @@
+      <RestaurantRegions regions={regions} />
+    ));
+
+    expect(getByText('서울')).not.toBeNull();
+    expect(getByText('대전')).not.toBeNull();
+    expect(getByText('대구')).not.toBeNull();
+    expect(getByText('부산')).not.toBeNull();
+    expect(getByText('광주')).not.toBeNull();
+    expect(getByText('강원도')).not.toBeNull();
+    expect(getByText('인천')).not.toBeNull();
```
`regions`를 이용해서 더 간단하게 표현해줄 수 있을 것 같아요. 그리고 지역이 하나만 더 추가되더라도 이쪽에 테스트를 수정해야하잖아요  하지만 fixture로 관리하면 fixture만 고쳐줘도 되서 fixture를 사용해서 표현하면 더 좋을 것 같아요

hannut91 reviewed (changes requested):
전체적으로 잘 적용하셨는데 TDD가 조금 아쉬워요. 코드를 작성하기 전에 테스트를 먼저 작성하면서 내가 구현해야할 것이 무엇인지 더 명확하게 하는 과정을 추가하면 좋을 것 같고, TDD과정에서 리팩터링이 많이 빠져있는 것 같아요. 리팩터링과정을 꼭 넣어서 해보세요. 검사하겠습니당

Kwakcena committed:
Refactor: 중복 코드 제거, beforeEach 사용

중복되는 테스트 코드를 fixtures의 데이터를 이용해 제거함.

beforeEach는 각각의 테스트 함수가 실행되기 전에 실행 할 내용을 담고있는
부분으로, given-when-then에서 given에 해당됨.

mockClear는 모의 함수 호출에 대한 모든 정보를 초기화 한다. 호출 정보는
.mock.calls와 .mock.results에 배열로 저장된다.

Kwakcena committed:
test(red): reducer setCategories test

레스토랑의 카테고리를 categories 배열에 넣어주는 테스트 작성

Kwakcena committed:
test(green): set categories 테스트 통과

카테고리 목록을 set 하는 테스트 통과

Kwakcena committed:
test(red): 카테고리 목록 보여주는 컴포넌트 작성

카테고리 목록을 화면에 보여주는 컴포넌트에 대해 테스트 함.

Kwakcena committed:
test(red): 테스트 수정

categories를 props로 넘겨주는 작업 필요

Kwakcena committed:
test(green): 하드코딩으로 통과

Kwakcena committed:
test(refactor): props를 사용하도록 리팩토링

들어오는 카테고리 목록을 전부 버튼으로 불러올 수 있도록 리팩토링

Kwakcena committed:
test(refactor): categories fixtures 추가

fixtures에 테스트용 categories 추가
버튼 타입인지 확인하는 테스트 추가

Kwakcena committed:
test(red): RestaurantCategoriesContainer 컴포넌트

레스토랑 카테고리 상태와 화면에 보여주는 RestaurantCategories 컴포넌트를
연결시켜주는 container 컴포넌트를 만들기 전 테스트 수행

Kwakcena committed:
test(red): 테스트 코드 수정

dispatch는 없어도 될 것 같음

Kwakcena committed:
test(green): fixtures를 사용해서 테스트 통과

Kwakcena committed:
test(refactor): useSelector 적용

Kwakcena committed:
test(red): App 컴포넌트에 적용

App 컴포넌트에서 이제 RestaurantCategoriesContainer 컴포넌트를
적용해본다.

Kwakcena committed:
test(green): 컨테이너 적용해서 통과

Kwakcena committed:
test(red): dispatch가 두번 불려야 됨

Kwakcena committed:
Refactor: 코드 수정

Kwakcena committed:
test(green): 리덕스 thunk 사용 전

사용 전에는 fixtures 의 데이터를 setCategories에 넣어 확인한다

Kwakcena committed:
test(green): api 서버를 사용해서 카테고리 가져오기

실제 api 서버를 이용해서 카테고리 목록을 가져온다.

Kwakcena committed:
Refactor: reducer 리팩토링

if문 제거 리팩토링

Kwakcena committed:
Refactor: api 함수 하나로 합치기

매개변수에 원하는 정보를 넣어주면 가져올 수 있도록 api의 함수를 하나로
합쳐 사용한다.

Kwakcena committed:
test(red): reducer test

이전의 state가 없는 경우 initialState를 사용한다는 테스트 추가

Kwakcena committed:
test(green): 초기 상태 사용 테스트 통과

Kwakcena committed:
test(red): 지역 클릭 시 체크 표시

해당 지역을 클릭하면 서울 -> 서울(V) 가 될 수 있도록 테스트함.

Kwakcena committed:
test(red): select region

지역 선택에 따라 체크 표시가 되는 테스트를 하기 전에 reducer로 선택된
지역을 갖게 되는 테스트를 해야 한다고 생각함

Kwakcena committed:
test(green): 선택된 지역을 region에 담는 테스트

선택한 지역을 region에 담는 테스트를 통과함.

초기 상태에도 region 변수를 만듬

Kwakcena committed:
test(red): 지역 선택 시 dispatch 호출 테스트

해당 지역을 선택하면 dispatch에 action이 전달되는 테스트를 진행

Kwakcena committed:
test(green): 선택한 region 담기, dispatch 호출

선택한 region을 담는 테스트와 dispatch가 호출되는 테스트를 통과한다.

Kwakcena committed:
test(Refactor): 테스트에서 actions 사용

Kwakcena committed:
test(refactor): 테스트 수정

RestaurantRegions 에서는 버튼 클릭 시 체크 표시가 되는 테스트를 할
필요가 없을 것 같다.

Kwakcena committed:
test(refactor): 초기값인 region을 갖도록 함

Kwakcena committed:
test(red): 선택한 분류를 카테고리에 담기

Kwakcena committed:
test(red): actions 추가, 코드 수정

dispatch를 mock 하는 코드로 수정

actions을 추가함. 현재 dispatch가 RestaurantCategories에 전달하여
호출하지 않아서 테스트 꺠짐

Kwakcena committed:
test(red): RestaurantCategories

버튼 클릭시 함수 호출 테스트

Kwakcena committed:
test(green): 클릭시 함수 호출 테스트 완료

Kwakcena committed:
test(green): RestaurantCategoriesContainer

컨테이너에서 핸들러 함수를 전달하여 dispatch가 호출되는지에 대한
테스트를 통과시킴.

Kwakcena committed:
test(red): 카테고리 선택에 대한 reducer 테스트

카테고리를 선택하면 해당 카테고리가 category 변수에 담기도록 한다.

Kwakcena committed:
test(green): reducer의 초기값에 category추가

이를 추가하고 action과 reducer를 정의하여 테스트 통과

Kwakcena committed:
오타 수정

Kwakcena committed:
Refactor: 테스트 코드에 category 초기값 추가

Kwakcena commented:
버튼 클릭시 화면에 체크 모양이 보이도록 하는 부분을 해결 못했어요. 클릭시 category와 region이 제대로 변하는 것 같지도 않네요... 테스트에서는 통과했는데 실제 console.log로 찍어보니 이전값이 찍히는데... 일단 한데까지만 제출하고 이어서 하겠습니다.

TDD 사이클을 지키는데 뭔가 어렵게 느껴지네요 ㅠ 

테스트의 순서를 어떻게 해야 할지 고민이에요.
container를 테스트 하다가 내부에서 쓰고있는 컴포넌트 테스트로 넘어가고, 부모 컴포넌트 -> 자식 컴포넌트 순서로 테스트를 진행해야 하는지, 내부의 자식 컴포넌트를 하나씩 테스트를 완료하고 바깥으로 나오는식으로 해야하는지 햇갈려요. 
그래서 그런지 테스트 순서가 굉장히 복잡해지더라구요.

hannut91 commented:
큰 문제에서 작은 문제로, 작은문제에서 큰 문제로 푸는 방법은 둘다 학습하셔야 합니다 ㅎㅎ 정확한것은 `어떻게 문제를 풀 것인가?`에 나오긴 하지만  큰 문제에서 작은 문제로 쪼개나갈 때는 실제로 뭐가 필요한지를 먼저 생각해놓고 풀 수 있다는 장점이 있는 것 같고, 작은 문제를 먼저 풀어서 큰 문제로 확장할 때는 문제 자체가 작기 때문에 단순해서 개발하기 쉽다는 장점이 있지만 실제로 어떻게 쓰일지에 대해서는 충분히 고려하기가 어렵다는 단점이 있어요. 이 과정은 자유롭게 왔다갔다 할 수 있어야 합니다.

See also
  - http://www.kyobobook.co.kr/product/detailViewKor.laf?barcode=9788981723422

hannut91 commented on src/reducer.test.js:
```diff
@@ -18,4 +20,13 @@ describe('reducer', () => {
     expect(state.regions).toHaveLength(1);
   });
+
+  it('set restaurant category into categorys', () => {
```
`categories`가 맞을 것 같아요!

hannut91 commented on src/reducer.test.js:
```diff
@@ -24,9 +24,10 @@ describe('reducer', () => {
 
     const state = reducer(initialState, setCategories(categories));
-    expect(state.categories).toHaveLength(1);
+    expect(state.categories).toHaveLength(2);
```
왜 2 인지 알기가 어려워요. 왜 3이 나 4는 안되나요? 이런 문자를 Magic Literal이라고 불러요. 이건 어떤가요?

```js
expect(state.categories).toEqual(categories);
```

See also
  - https://refactoring.com/catalog/replaceMagicLiteral.html

hannut91 commented on src/RestaurantCategories.test.jsx:
```diff
@@ -4,15 +4,17 @@ import { render } from '@testing-library/react';
 
-    expect(getByText('한식')).not.toBeNull();
+    categories.forEach(({ name }) => {
+      expect(getByText(name)).not.toBeNull();
+      expect(getByText(name)).toHaveAttribute('type', 'button');
```
찾은 요소가 존재하지 않는지 확인하고 속성이 버튼인지 테스트하는 것을 작성하신 것 같아요. 속성을 확인하는 것만으로도 해당 요소가 존재하는걸 확인할 수 있으니 속성만 확인하는건 어떨까요?

hannut91 commented on src/RestaurantCategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+    });
+  });
+
+  it('선택한 지역을 category에 담는다.', () => {
```
선택한 지역이 아니라 선택한 분류인 것 같아요

hannut91 commented on src/RestaurantCategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+      <RestaurantCategoriesContainer />
+    ));
+
+    fireEvent.click(getByText('한식'));
```
여기도 `category`를 이요해서 표현해주면 좋을 것 같아요

hannut91 commented on src/RestaurantRegionsContainer.test.jsx:
```diff
@@ -0,0 +1,46 @@
+      <RestaurantRegionsContainer />
+    ));
+
+    fireEvent.click(getByText('광주'));
```
여기도 `광주`대신에  `region`으로 표현해주면 좋을 것 같아요

hannut91 commented on src/reducer.js:
```diff
@@ -0,0 +1,33 @@
+};
+
+export default function reducer(state = initialState, action = {}) {
+  const reducers = {
+    setRegions: ({ regions }) => ({
+      ...state,
+      regions,
+    }),
+    setCategories: ({ categories }) => ({
+      ...state,
+      categories,
+    }),
+    selectRegion: ({ region }) => ({
+      ...state,
+      region,
+    }),
+    selectCategory: ({ category }) => ({
+      ...state,
+      category,
+    }),
+  };
```
이러면 reducer가 실행될 때 마다 계속 만들어 질 것 같아요. 함수 바깥에 선언하는게 좋을 것 같아요

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,54 @@
+
+    const state = reducer(initialState, setRegions(regions));
+
+    expect(state.regions).toHaveLength(1);
```
`expect(state.regions).toEqual(regions)`는 어떨까요?

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,54 @@
+    expect(state.categories).toHaveLength(2);
+  });
+
+  it('select region', () => {
+    const region = '서울';
+    const state = reducer(initialState, selectRegion(region));
+    expect(state.region).toBe('서울');
+  });
+
+  it('select category', () => {
+    const category = '한식';
+    const state = reducer(initialState, selectCategory(category));
+    expect(state.category).toBe('한식');
+  });
```
여기도 마찬가지로 Magic literal을 없애면 좋을 것 같아요

Kwakcena commented:
리팩토링 해야 할게 이리도 많았군요... 죄송합니다.😭
